Fixes #833

nilay77 opened the pull request:
- Added a class variable `size` in Memory Store to track the number of triples in the store. This allows returning the number of triples in the store in O(1) instead of O(n)

ashleysommer commented on rdflib/plugins/memory.py:
```diff
@@ -49,6 +52,15 @@ def add(self, triple, context, quoted=False):
+
+        try:
+            # Triple already in store
+            if(spo[subject][predicate][object]==1):
```
Like the fast key tests below, you can replace this without an if statement like so:
```python
try:
    # Triple already in store
    _ = spo[subject][predicate][object]
except:
    # New triple being added
    self.__size += 1
```

ashleysommer replied:
Don't need to check if the value `== 1`, because when the triple is removed, the whole key+value is removed from the corresponding dict.

nilay77 replied:
Yes, that makes sense. I'll change it.

ashleysommer commented on rdflib/plugins/memory.py:
```diff
@@ -49,6 +52,15 @@ def add(self, triple, context, quoted=False):
+        try:
+            # Triple already in store
+            if(spo[subject][predicate][object]==1):
+                pass
```
I think you could actually `return` here, rather than `pass`.
After spending the cpu time to check if the triple already exists in the index, we should at least take advantage of that answer and exit the method early if we know it exists.
(This assumes that if it exists in the `spo` index, then it also exists in the `pos` and the `osp` indexes)

nilay77 replied:
Your assumption makes sense, I did not think of this. I'll update the code. Thanks!

nilay77 commented:
- Added functionality to return 1 on the addition of a new triple in Memory and IOMemory store, and 0 otherwise.
- Modified addN method from store.py to return the number of new triples/quads added. 
- Modified add method from Graph to return 1 on the addition of a new triple, and 0 otherwise.
- Modified addN method from Graph to return the number of new triples/quads added.
- Added test cases for add and addN method of Graph.

nilay77 commented:
- Fixes #833: Added functionality to return the count of the number of triples added from Graph.addN

mwatts15 commented on test/test_issue833.py:
```diff
@@ -0,0 +1,61 @@
+        self.quads.append((bob, hates, pizza, self.graph))
+
+    def testAddN(self):
+        asserte = self.assertEqual
```
you don't really need to define `asserte`. It's defined in other places because it's used so many times in the same method.

mwatts15 commented on test/test_issue833.py:
```diff
@@ -0,0 +1,61 @@
+
+    def testAddN(self):
+        asserte = self.assertEqual
+        tarek = self.tarek
+        bob = self.bob
+        likes = self.likes
+        hates = self.hates
+        pizza = self.pizza
+        cheese = self.cheese
```
Recommend to just define these on the module level -- less code and no reason not to.

(Don't ask me why it's done that way in other tests. IDK. It's kinda dumb.)

nilay77 replied:
Thanks for all the suggested changes. I have added them, please have a look.

mwatts15 replied:
Looks OK to me.

(I was actually recommending to define the `URIRef`s _outside_ of the class, like on line 7.)

tgbugs commented on rdflib/plugins/memory.py:
```diff
@@ -38,15 +38,29 @@ def __init__(self, configuration=None, identifier=None):
         spo = self.__spo
+        try:
+            # Triple already in store
+            _ = spo[subject][predicate][object]
```
What is the performance impact here?

nilay77 replied:
As far as I [know](https://stackoverflow.com/questions/1963507/time-complexity-of-accessing-a-python-dict), on an average, it should be O(1).

tgbugs commented on rdflib/plugins/memory.py:
```diff
@@ -38,15 +38,29 @@ def __init__(self, configuration=None, identifier=None):
         self.__prefix = {}
 
+        # number of triples in the store
+        self.__size = 0
```
Are there cases where this variable could get out of sync with the actual number in the store?

nilay77 replied:
I do not think so. The number of triples only changes when the add or remove function is called, and in both the cases this variable is updated accordingly.

tgbugs commented on rdflib/store.py:
```diff
@@ -226,14 +226,18 @@ def addN(self, quads):
+            assert c is not None, \
+                "Context associated with %s %s %s is None!" % (s, p, o)
+            added = self.add((s, p, o), c)
+            if (isinstance(added,int)):
```
Stores that do not support count triples should not be forced to deal with this branch. Should probably dispatch on whether or not the store supports returning counts.

nilay77 replied:
Stores that do not support count triples would not return anything from their `add `method, in which case `None` will be returned from this (i.e. `addN`) method. I could not understand how this would be an issue?

mwatts15 replied:
Since there hasn't been any movement on this: I was advocating for an explicit tag on stores to know if a count is returned, but at this point, we weren't using the return value of add/addN for anything, so until another feature comes along, this should be fine.

As far as "dealing with the branch", I assume you're thinking of performance based on your general comment, but this isn't a tight loop since we're making a function call, so there's barely in additional cost here.

tgbugs commented on rdflib/plugins/memory.py:
```diff
@@ -249,6 +262,7 @@ def namespaces(self):
             yield prefix, namespace
 
     def add(self, triple, context, quoted=False):
+        count_before = self.__len__()
```
Why not use `len(self)`?

nilay77 replied:
That should work too. I thought `self.__len__()` would be a direct call and `len(self)` would be an indirect call, thus I preferred the former. Should I change it to `len(self)`?

nicholascar commented:
Hi @nilay77 sorry but this PR is now outdated due to changes in the default Store and the names of Stores. We've introduced a new, more efficient, Store and renamed the old Memory Store to SimpleMemory.

Secondly, while I don't disagree that this PR will change `len()` from O(n) to O(1), why is this particularly valuable? Even for the largest graphs, an O(n) count isn't going to kill any computers and, with this code, you've now added the additional variable and counting operations for every add/addN/remove call.

Finally, this store can't really be used for much - it's not the default - since it's not capable of some Store functions needed for many RDF formats.

Won't merge and closing, sorry!
